cli/flox/src/commands/environment.rs

```diff
@@ -1380,20 +1386,104 @@ impl Pull {
 
         let pointer_content =
             serde_json::to_string_pretty(&pointer).context("Could not serialize pointer")?;
-        let pointer_path = dot_flox_path.join(ENVIRONMENT_POINTER_FILENAME);
 
-        fs::create_dir_all(&dot_flox_path).context("Could not create .flox/ directory")?;
+        let dot_flox_parent = dot_flox_path
+            .parent()
+            .context("Could not get .flox/ parent")?;
+        fs::create_dir_all(dot_flox_parent).context("Could not create .flox/ parent directory")?;
+
+        // Pulls the environment into a temp directory which is later renamed to `.flox`.
+        // We do this to avoid populating the floxmeta branch `owner/name.encode(path to .flox)`
+        // in case building fails or the user aborts the fixup.
+        // The branch will be adjusted when the environment is opened the next time
+        // by the `ensure_branch` routine.
+        let temp_dot_flox_dir = tempfile::TempDir::with_prefix_in(DOT_FLOX, dot_flox_parent)
+            .context("Could not create temporary directory for cloning environment")?;
+
+        let pointer_path = temp_dot_flox_dir.path().join(ENVIRONMENT_POINTER_FILENAME);
         fs::write(pointer_path, pointer_content).context("Could not write pointer")?;
 
-        let result =
-            ManagedEnvironment::open(flox, pointer, &dot_flox_path).map_err(Self::convert_error);
+        let mut env = {
+            let result = Dialog {
+                message,
+                help_message: None,
+                typed: Spinner::new(|| ManagedEnvironment::open(flox, pointer, &temp_dot_flox_dir)),
+            }
+            .spin()
+            .map_err(Self::convert_error);
+
+            match result {
+                Err(err) => {
+                    fs::remove_dir_all(&temp_dot_flox_dir)
+                        .context("Could not clean up .flox/ directory")?;
+                    Err(err)?
+                },
+                Ok(env) => env,
+            }
+        };
+
+        let result = Dialog {
+            message,
+            help_message: None,
+            typed: Spinner::new(|| env.build(flox)),
+        }
+        .spin();
+
         match result {
-            Err(err) => {
-                fs::remove_dir_all(dot_flox_path).context("Could not clean up .flox/ directory")?;
-                Err(err)?;
+            Ok(_) => {
+                fs::rename(temp_dot_flox_dir, dot_flox_path)
+                    .context("Could not move .flox/ directory")?;
+            },
+            Err(
+                e @ EnvironmentError2::Core(CoreEnvironmentError::LockedManifest(
+                    LockedManifestError::BuildEnv(CallPkgDbError::PkgDbError(PkgDbError {
+                        exit_code: 123,
```

Comment:
nonblocking: maybe make this a constant in Rust, so if we change numbers in C++ we have one place to look in Rust?